### DISCUSSION TITLE
Possible to change default cart behaviour regarding coupons

### DISCUSSION THREAD

**Discussion by coderreco** [swellstores/community#363] [Help] [answered]
Is it possible to update the settings for how coupons are handled in the cart?

A few examples:
-When using the ***Product or category*** discount type, can we exclude the discounted product from determining if the cart is eligible? Example: If _Discount A_ offers 50% off of _Product X_ with a minimum subtotal of $150, the item cost of _Product X_ does not count towards the minimum subtotal.
-Or if easier, globally distinguish whether or not a product can be counted towards discount eligibility from the product model itself?
-Not allowing a coupon to be applied to a cart if the cart doesn't yet meet the eligibility conditions. Current behaviour is it will add the coupon to the cart, but will not apply the discount until conditions are met.
-If cart meets coupon conditions and coupon has been applied, then shopper removes or changes items which removes eligibility for the coupon, have the option to automatically remove coupon.
-The current auto "Add product to cart when applied" ability can cause some problems, particularly with Promotions since they automatically applied to the cart. If the promo or coupon is applied to the cart, the product is impossible to delete because the coupon/promo automatically re-adds it. In my testing once a promotion was applied, even if the cart no longer met the eligibility requirements the promotion would not automatically delete itself from the cart. Therefore making it impossible for the shopper to delete the item. Since promotions are automatic there are no frontend-api methods to remove them, I'm sure I could use the backend-api and do a `swell.put` on the cart and remove the promotion but this doesn't seem like the best way to handle this.

Thanks in advance!

**Comment by sags95** [comment, marked as answer]
@coderreco For most of these unfortunately it is not possible to change the actual logic for implementing discounts beyond what is available in the actual coupons/promotions model. What we have seen in these cases is developing a custom discount service that can be applied using the backend api (Swell support custom discounts). Using custom fields on coupons/promotions in combination with another service can likely accomplish these requests.

For that last part, I will be sure to file this with the team as this behaviour is likely unintended.

**Reply by coderreco** [reply]
Thanks @sags95.

Do you have any examples of how others have implemented custom discounts? I have begun work on my own custom validation service that runs before applying coupons to the cart, but would love more context on how Swell supports custom discounts.

**Reply by logeshswell** [reply]
@coderreco You can refer to this guide to implement custom discount service using real-time order webhooks.
https://developers.swell.is/guides/core-concepts/webhook-basics#real-time-order-webhooks
